### ISSUE TITLE
Unable to pass wildcard when searching for JWST data

### ISSUE THREAD

**Issue by bhilbert4** [astropy/pyvo#435]
Hello. I'm trying to use pyvo to search MAST for the rootnames associated with a given program number and instrument. With the recent update to MAST, there are now potentially multiple instrument values associated with a given program. I was hoping to be able to add a wildcard to the instrument name, so that all of the potential instrument values would be covered by the single query. However, when I try including the wildcard, the code crashes. I spoke with David Rodriguez, and he mentioned that the wildcard does work in TAP. Is pyvo expecting some wildcard character other than a %? Below is my query and the traceback.

thanks,
Bryan

```
import pyvo as vo

instrument = 'miri%'
proposal = '2311'

tap_service = vo.dal.TAPService("http://vao.stsci.edu/caomtap/tapservice.aspx")
tap_results = tap_service.search(f"select observationID from dbo.CaomObservation where collection='JWST' and maxLevel=2 and insName like 'miri%' and prpID='{int(proposal)}'")
prop_table = tap_results.to_table()
rootnames = prop_table['observationID'].data
```
```
---------------------------------------------------------------------------
ConnectionResetError                      Traceback (most recent call last)
~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    698             # Make the request on the httplib connection object.
--> 699             httplib_response = self._make_request(
    700                 conn,

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    444                     # Otherwise it looks like a bug in the code.
--> 445                     six.raise_from(e, None)
    446         except (SocketTimeout, BaseSSLError, SocketError) as e:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/packages/six.py in raise_from(value, from_value)

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    439                 try:
--> 440                     httplib_response = conn.getresponse()
    441                 except BaseException as e:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/http/client.py in getresponse(self)
   1346             try:
-> 1347                 response.begin()
   1348             except ConnectionError:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/http/client.py in begin(self)
    306         while True:
--> 307             version, status, reason = self._read_status()
    308             if status != CONTINUE:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/http/client.py in _read_status(self)
    267     def _read_status(self):
--> 268         line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
    269         if len(line) > _MAXLINE:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/socket.py in readinto(self, b)
    668             try:
--> 669                 return self._sock.recv_into(b)
    670             except timeout:

ConnectionResetError: [Errno 54] Connection reset by peer

During handling of the above exception, another exception occurred:

ProtocolError                             Traceback (most recent call last)
~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    438             if not chunked:
--> 439                 resp = conn.urlopen(
    440                     method=request.method,

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    754 
--> 755             retries = retries.increment(
    756                 method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/util/retry.py in increment(self, method, url, response, error, _pool, _stacktrace)
    531             if read is False or not self._is_method_retryable(method):
--> 532                 raise six.reraise(type(error), error, _stacktrace)
    533             elif read is not None:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/packages/six.py in reraise(tp, value, tb)
    768             if value.__traceback__ is not tb:
--> 769                 raise value.with_traceback(tb)
    770             raise value

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    698             # Make the request on the httplib connection object.
--> 699             httplib_response = self._make_request(
    700                 conn,

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    444                     # Otherwise it looks like a bug in the code.
--> 445                     six.raise_from(e, None)
    446         except (SocketTimeout, BaseSSLError, SocketError) as e:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/packages/six.py in raise_from(value, from_value)

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    439                 try:
--> 440                     httplib_response = conn.getresponse()
    441                 except BaseException as e:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/http/client.py in getresponse(self)
   1346             try:
-> 1347                 response.begin()
   1348             except ConnectionError:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/http/client.py in begin(self)
    306         while True:
--> 307             version, status, reason = self._read_status()
    308             if status != CONTINUE:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/http/client.py in _read_status(self)
    267     def _read_status(self):
--> 268         line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
    269         if len(line) > _MAXLINE:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/socket.py in readinto(self, b)
    668             try:
--> 669                 return self._sock.recv_into(b)
    670             except timeout:

ProtocolError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

ConnectionError                           Traceback (most recent call last)
~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/dal/query.py in execute_votable(self, post)
    241         try:
--> 242             return votableparse(self.execute_stream(post=post).read)
    243         except Exception as e:

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/dal/tap.py in execute_stream(self, post)
    925 
--> 926         return super().execute_stream(post=post)
    927 

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/utils/decorators.py in wrapper(*args, **kwargs)
      8     def wrapper(*args, **kwargs):
----> 9         raw = func(*args, **kwargs)
     10         raw.read = partial(raw.read, decode_content=True)

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/dal/query.py in execute_stream(self, post)
    190         """
--> 191         response = self.submit(post=post)
    192 

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/dal/tap.py in submit(self, post)
    958 
--> 959         response = self._session.post(
    960             url, data=self, stream=True, files=files)

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/requests/sessions.py in post(self, url, data, json, **kwargs)
    589 
--> 590         return self.request('POST', url, data=data, json=json, **kwargs)
    591 

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/requests/sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    541         send_kwargs.update(settings)
--> 542         resp = self.send(prep, **send_kwargs)
    543 

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/requests/sessions.py in send(self, request, **kwargs)
    654         # Send the request
--> 655         r = adapter.send(request, **kwargs)
    656 

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    497         except (ProtocolError, socket.error) as err:
--> 498             raise ConnectionError(err, request=request)
    499 

ConnectionError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

DALFormatError                            Traceback (most recent call last)
<ipython-input-60-e7c76dfb8362> in <module>
      1 tap_service = vo.dal.TAPService("http://vao.stsci.edu/caomtap/tapservice.aspx")
----> 2 tap_results = tap_service.search(f"select observationID from dbo.CaomObservation where collection='JWST' and maxLevel=2 and insName like 'miri%' and prpID='{int(proposal)}'")
      3 prop_table = tap_results.to_table()
      4 rootnames = prop_table['observationID'].data

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/dal/tap.py in run_sync(self, query, language, maxrec, uploads, **keywords)
    244         TAPResults
    245         """
--> 246         return self.create_query(
    247             query, language=language, maxrec=maxrec, uploads=uploads,
    248             **keywords).execute()

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/dal/tap.py in execute(self)
    940            for errors parsing the VOTable response
    941         """
--> 942         return TAPResults(self.execute_votable(), url=self.queryurl, session=self._session)
    943 
    944     def submit(self, post=False):

~/miniconda3/envs/jwql-py3.8/lib/python3.8/site-packages/pyvo/dal/query.py in execute_votable(self, post)
    243         except Exception as e:
    244             self.raise_if_error()
--> 245             raise DALFormatError(e, self.queryurl)
    246 
    247     def raise_if_error(self):

DALFormatError: ConnectionError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))
```

**Comment by bsipocz** [comment]
What is your pyvo version? your snippet works for me with v1.4.1

**Comment by bhilbert4** [comment]
Looks like I'm using v1.3. I'll try updating.

**Comment by bhilbert4** [comment]
I'm getting the same result in v1.4 and v1.4.1. Maybe something to do with the STScI VPN? I'm not sure what would be causing the connection to be reset.

**Comment by tomdonaldson** [comment]
@bhilbert4 Yes, that matches the symptoms of our firewalls rejecting TAP queries. It's worth noting that the rejections can happen both ways, so if you're on site (even eduroam) or on VPN, TAP queries to other archives can be rejected.  

I will follow up with our security team to try to allow that traffic again, at least to MAST services.

**Comment by tomdonaldson** [comment]
@bhilbert4 Changing the service URL to use `https` will probably get you through.  Could you please try with `https://vao.stsci.edu/caomtap/querymanager.html`?

**Comment by bsipocz** [comment]
Seeing that this is an ST and firewall issue and there isn't really much to do about in pyvo, I go ahead and close the issue. Feel free to continue the discussion if needed.

**Comment by bhilbert4** [comment]
Thanks Tom. I updated to https and the query went through. I'll update JWQL to use that as well, although JWQL hasn't had any problems with this error so far.

**Comment by theresadower** [comment]
Thank you @tomdonaldson  for following up before I saw this!

In an attempt to sidestep this ongoing problem with the ST network (and save some time in forwarding hops), I recently changed our IVOA registry to announce https endpoints for our TAP services. This will keep new discoveries of the service from running into the issue, but unfortunately there are many pieces of existing software out in the wild with their own saved copies of the endpoints to update. As a note, this will also apply to our other TAP services for HSC, PanSTARRS, and the registry itself.
